### PR TITLE
Allow configuring a command to be automatically run when disconnecting from VPN (take 2)

### PR DESCRIPTION
**Pull request by jherland**
I configure my SSH with ControlMaster connections that must be closed when I disconnect from VPN, otherwise they are left stale and existing and future SSH session are left hanging/wedged.

To accomodate this, the last patch in this series teaches openconnect-sso automatically run a configured command on VPN disconnection. In my case, I run a shell script that does `ssh -O exit ...` on my connections, although this doesn't matter from openconnect-sso's POV.

Otherwise the other patches are only tangentially related:
- Patch#1 fixes an obvious bug in the `develop` branch.
- Patch#2 stops openconnect-sso from crashing when the config file cannot be written. It allows me to control the openconnect-sso configuration via [home-manager](https://github.com/nix-community/home-manager) (which stores a symlink to a read-only file in the Nix store).
- Patch#3 moves code from the async `_run()` into the sync `run()`. Specifically it moves the running of the `openconnect` subprocess which does not benefit from being async. This also fixes the control flow when aborting `openconnect` with `Ctrl+C`.

This supersedes https://github.com/vlaci/openconnect-sso/pull/32, but is based on `develop` instead of `master`, as I found it easier to get this done on top of those changes.

### CODE REVIEW

**Comment by vlaci** [comment]
I'll check it out tomorrow.

File: openconnect_sso/config.py [hunk 1]
```diff
@@ -62,6 +62,7 @@ class HostProfile(ConfigNode):
     address = attr.ib(converter=str)
     user_group = attr.ib(converter=str)
     name = attr.ib(converter=str)  # authgroup
+    on_disconnect = attr.ib(converter=str)
```
Comment (vlaci) [inline openconnect_sso/config.py:65]: If you'd like to persist this option, please put this attribute in the `Config` class instead. This is not relates to the connection details.

File: openconnect_sso/app.py [hunk 2]
```diff
@@ -179,3 +181,10 @@ def run_openconnect(auth_info, host, proxy, args):
 
     logger.debug("Starting OpenConnect", command_line=command_line)
     return subprocess.run(command_line).returncode
+
+
+def handle_disconnect(command):
+    if command:
+        logger.info(f"Running {command!r} on shutdown...")
```
Comment (vlaci) [inline openconnect_sso/app.py:188]: Please use keyword arguments to template log messages, e.g `logger.info("Running command on disconnect", command_line=command)

File: openconnect_sso/cli.py [hunk 3]
```diff
@@ -60,6 +60,12 @@ def create_argparser():
         default="",
     )
 
+    auth_settings.add_argument(
```
Comment (vlaci) [inline openconnect_sso/cli.py:63]: This setting is not related to authentication, please use the global parser object for that.

File: openconnect_sso/app.py [hunk 4]
```diff
@@ -179,3 +181,10 @@ def run_openconnect(auth_info, host, proxy, args):
...
+        argv = [str(Path(command).expanduser())]
+        return subprocess.run(argv, timeout=5).returncode
```
Comment (vlaci) [inline openconnect_sso/app.py:190]: IMHO this command does not support arguments right now, is that right? If you set `shell=True` in `subprocess.run` then you could use a command line with arguments. (or you can split the command line with `shlex.quote` instead)

File: openconnect_sso/app.py [hunk 5]
```diff
@@ -28,17 +29,50 @@ def run(args):
     try:
         if os.name == "nt":
             asyncio.set_event_loop(asyncio.ProactorEventLoop())
-        return asyncio.get_event_loop().run_until_complete(_run(args))
+        auth_response, selected_profile = asyncio.get_event_loop().run_until_complete(
+            _run(args)
+        )
     except KeyboardInterrupt:
         logger.warn("CTRL-C pressed, exiting")
+        return 1
```
Comment (vlaci) [inline openconnect_sso/app.py:37]: Previously CTRL-C was not regarded as an error, it exited gracefully. I can see the value of returning an error code for it. AFAIK it should be 130 (`-2` in twos complement, or 256+signal.SIGINT casted to a byte) for CTRL-C

**Review by vlaci — CHANGES_REQUESTED** [review]

File: openconnect_sso/config.py (hunk 1 above)
Comment (jherland) [reply openconnect_sso/config.py:65]: Agreed. Will move.

File: openconnect_sso/app.py (hunk 2 above)
Comment (jherland) [reply openconnect_sso/app.py:188]: Will do. I'm new to `structlog`. Looks very nice.

File: openconnect_sso/cli.py (hunk 3 above)
Comment (jherland) [reply openconnect_sso/cli.py:63]: Ack. Will do.

File: openconnect_sso/app.py (hunk 4 above)
Comment (jherland) [reply openconnect_sso/app.py:190]: Correct. Didn't need arguments in my specific use case, but it's certainly needed in a more general version. Since we're going via a string in the config file, I'll use `shell=True` here. Although I'm usually not a fan of `shell=True`, we should be fine as all the input ultimately comes from the (trusted) user. Alternatively, we could use a list of string args in `config.toml`, but that doesn't feel very natural.

File: openconnect_sso/app.py (hunk 5 above)
Comment (jherland) [reply openconnect_sso/app.py:37]: My thinking here was that if you abort with CTRL-C during the initial part of the script (before we launch openconnect), then a non-zero exit code is warranted, because one cannot argue that the program has completed successfully in any way, when the user aborted authentication and we never got around to even starting `openconnect`.

Conversely, once `openconnect` is running, pressing CTRL-C is a natural way to disconnect the VPN session (is there an easier way for the user to signal the intent to disconnect?), so CTRL-C then should not by regarded as an error. You'll see this towards the end of `run()` where I handle `KeyboardInterrupt` with `return 0`.

That said, I will change the exit code in the first case from `1` to `130`.

**Comment by jherland** [comment]
Pushed updated version with requested changes.

**Review by vlaci — APPROVED** [review]

**Comment by vlaci** [comment]
Thank you for your contribution. Will try to publish a new release this weekend.
